### PR TITLE
Add Player Client Options API

### PR DESCRIPTION
**Pull request by MiniDigger**
https://streamable.com/uimm5

```java
    @EventHandler
    public void onTest(PlayerClientOptionsChangeEvent event) {
        Bukkit.broadcastMessage(event.getPlayer().getDisplayName() + " just changed his settings!");
        if (event.hasLocaleChanged()) {
            Bukkit.broadcastMessage("Locale changed from " + event.getPlayer().getClientOption(ClientOption.LOCALE) + " to " + event.getLocale());
        }
        if (event.hasChatVisibilityChanged()) {
            Bukkit.broadcastMessage("ChatVisibility changed from " + event.getPlayer().getClientOption(ClientOption.CHAT_VISIBILITY) + " to " + event.getChatVisibility());
        }
        if (event.hasMainHandChanged()) {
            Bukkit.broadcastMessage("MainHand changed from " + event.getPlayer().getClientOption(ClientOption.MAIN_HAND) + " to " + event.getMainHand());
        }
        if (event.hasSkinPartsChanged()) {
            Bukkit.broadcastMessage("SkinParts changed from " + event.getPlayer().getClientOption(ClientOption.SKIN_PARTS) + " to " + event.getSkinParts());
        }
        if (event.hasViewDistanceChanged()) {
            Bukkit.broadcastMessage("ViewDistance changed from " + event.getPlayer().getClientOption(ClientOption.VIEW_DISTANCE) + " to " + event.getViewDistance());
        }
        if (event.hasChatColorsEnabledChanged()) {
            Bukkit.broadcastMessage("ChatColors changed from " + event.getPlayer().getClientOption(ClientOption.CHAT_COLORS_ENABLED) + " to " + event.hasChatColorsEnabled());
        }
    }
```

### CODE REVIEW

**Comment by MiniDigger** [comment]
ready for review

File: Spigot-API-Patches/0191-Add-PlayerSettingsChangeEvent.patch [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
++    public int getSkinparts() {
++        return skinparts;
++    }
++
++    @NotNull
++    public MainHand getMainHand() {
++        return mainHand;
++    }
++
++    public boolean hasCapeEnabled() {
```
Comment (Spottedleaf) [inline Spigot-API-Patches/0191-Add-PlayerSettingsChangeEvent.patch]: let's not make the impl detail of skins api

Comment (MiniDigger) [reply Spigot-API-Patches/0191-Add-PlayerSettingsChangeEvent.patch]: Mmh, ok, I will let the event return a new SkinParts API class with an impl that handles that.

Comment (MiniDigger) [reply Spigot-API-Patches/0191-Add-PlayerSettingsChangeEvent.patch]: done

**Comment by Phoenix616** [comment]
Would it be possible to add a way of getting which setting(s) changed in the event?

**Comment by electronicboy** [comment]
obfhelpers might be nice; Also could implement ^ by having a set of enums values with what's changed, saying that, not really sure that's needed, however, but would be a nicety

Might potentially be nice to move the skinparts stuff to its own patch and exposing that or something

**Comment by MiniDigger** [comment]
ready for review again.
addressed your comments @electronicboy @Phoenix616

File: Spigot-API-Patches/0190-Add-Player-Settings-API.patch [hunk 2]
```diff
@@ -20,16 +20,16 @@ index 00000000..27aed60d
 +
 +public class SettingType<T> {
```
Comment (yannicklamprecht) [inline Spigot-API-Patches/0190-Add-Player-Settings-API.patch]: final. we do not want to have inheritance here.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

**Review by Trigary — COMMENTED** [review]
I think it would be nice to be able to get both the old and the new values from the `PlayerSettingsChangeEvent` instance. Sure, the event's implementation demonstrates how to get both, but I don't think that's self-explanatory.

**Comment by electronicboy** [comment]
Just cludges the API up needlessly, the existing design of the event makes it trivial to get the new/old values pretty easily; The design of this event follows pretty standard design around the entire API, events fire before state modification

**Comment by Phoenix616** [comment]
Yeah I agree that cluttering up the event is a bit pointless if you can easily get the old value via the Player instance.

**Comment by MiniDigger** [comment]
final + squashed

**Comment by Trigary** [comment]
I kind of forgot that almost all events in the Bukkit API work like that, as electroniccat said:

> events fire before state modification

I guess I'm just being dumm as usual

**Comment by MiniDigger** [comment]
nobody saw that failed rebase.

its rebased now.

File: Spigot-API-Patches/0190-Add-Player-Settings-API.patch [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    @NotNull
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++
++    public enum ChatVisibility {
```
Comment (electronicboy) [inline Spigot-API-Patches/0190-Add-Player-Settings-API.patch]: This should probably be in SettingsType

File: Spigot-API-Patches/0190-Add-Player-Settings-API.patch [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+@@ -0,0 +1,28 @@
++package com.destroystokyo.paper;
++
++import com.destroystokyo.paper.event.player.PlayerSettingsChangeEvent.ChatVisibility;
++
++import org.jetbrains.annotations.NotNull;
++
++import org.bukkit.inventory.MainHand;
++
++public final class SettingType<T> {
```
Comment (electronicboy) [inline Spigot-API-Patches/0190-Add-Player-Settings-API.patch]: Only concern here is the ambiguity of the class name

File: Spigot-Server-Patches/0435-Implement-Player-Settings-API.patch [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
++import com.destroystokyo.paper.event.player.PlayerSettingsChangeEvent;
++import com.destroystokyo.paper.event.player.PlayerSettingsChangeEvent.ChatVisibility;
+ import com.destroystokyo.paper.profile.CraftPlayerProfile;
+ import com.destroystokyo.paper.profile.PlayerProfile;
+ import com.google.common.base.Preconditions;
+@@ -36,6 +40,7 @@ import net.minecraft.server.BlockPosition;
+ import net.minecraft.server.ChatComponentText;
+ import net.minecraft.server.Container;
+ import net.minecraft.server.Entity;
++import net.minecraft.server.EntityHuman;
```
Comment (electronicboy) [inline Spigot-Server-Patches/0435-Implement-Player-Settings-API.patch]: this still being used?

Comment (MiniDigger) [reply Spigot-Server-Patches/0435-Implement-Player-Settings-API.patch]: yes, to access a datawatcher

**Review by electronicboy — APPROVED** [review]
potentially missing a few // Paper comments, but otherwise looks fine
nitpick would be nice to be resolved before pull or later on would be the obfhelpers, we generally place the helper above the method, this just helps clean up the flow for merging

File: Spigot-Server-Patches/0435-Implement-Player-Client-Options-API.patch [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+ import java.util.Collection;
+ import java.util.Deque; // Paper
+@@ -16,7 +18,6 @@ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+ 
+ // CraftBukkit start
+-import com.google.common.base.Preconditions;
+ import org.bukkit.Bukkit;
+ import org.bukkit.GameMode;
+ import org.bukkit.Location;
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0435-Implement-Player-Client-Options-API.patch]: you totally busted the imports, revert

**Comment by MiniDigger** [comment]
@electronicboy you mean the obf helpers in PacketPlayInSettings? 
gonna take care of those and the imports later

File: Spigot-Server-Patches/0435-Implement-Player-Client-Options-API.patch (hunk 6 above)
Comment (MiniDigger) [reply Spigot-Server-Patches/0435-Implement-Player-Client-Options-API.patch]: done

**Comment by MiniDigger** [comment]
Resolved the issues

**Comment by MiniDigger** [comment]
Rebased, ready to merge

**Review by aikar — APPROVED** [review]
Looks good to me. sorry for delay on review.

**Review by zachbr — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Review by yannicklamprecht — APPROVED** [review]

**Review by Redned235 — APPROVED** [review]

**Review by Draycia — APPROVED** [review]

**Pull request merged**